src/Symfony/Component/TypeInfo/Type.php

```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
 {
     use TypeFactoryTrait;
 
-    abstract public function getBaseType(): BuiltinType|ObjectType;
-
     /**
-     * @param TypeIdentifier|class-string $subject
+     * Traverses and yields the whole type inheritance graph.
+     *
+     * For example traversing "string|\Traversable<int>" will yield
+     * - string
+     * - int
+     * - \Traversable<int>
+     * - string|\Traversable<int>
+     *
+     * @return iterable<self>
      */
-    abstract public function isA(TypeIdentifier|string $subject): bool;
+    public function traverse(): iterable
+    {
+        if ($this instanceof CompositeTypeInterface) {
+            foreach ($this->getTypes() as $type) {
+                yield from $type->traverse();
+            }
+        }
+
+        if ($this instanceof WrappingTypeInterface) {
+            yield from $this->getWrappedType()->traverse();
+        }
 
-    abstract public function asNonNullable(): self;
+        yield $this;
+    }
 
     /**
-     * @param callable(Type): bool $callable
+     * Tells if the type or one of its component (@see traverse()) is what is expected.
+     *
+     * @param TypeIdentifier|class-string $expected
      */
-    public function is(callable $callable): bool
+    public function isA(TypeIdentifier|string $expected): bool
     {
-        return $callable($this);
+        foreach ($this->traverse() as $type) {
+            if ($type !== $this && $type->isA($expected)) {
```

Comment:
Wouldn't that perform multiple traversals of parts of the object graph when one of the nested type is itself a composite or a wrapping type, and this type does not override `isA` ? `traverse` does not yield only leaf nodes